db.go

```diff
@@ -111,13 +95,38 @@ func (db *DB) doCompaction() {
 
 func (db *DB) checkAndTriggerCompaction() bool {
 	readyToExit := true
-	for idx, level := range db.levels {
-		if len(level.sstables) > maxLevelSSTables[idx] {
+
+	if db.needL0Compaction() {
+		db.compactionChan <- 0
+		readyToExit = false
+	}
+
+	if !db.needLevelNCompactions() {
+		readyToExit = false
+	}
+
+	return readyToExit
+}
+
+func (db *DB) needL0Compaction() bool {
+	return len(db.levels[0].sstables) > l0Capacity
+}
+
+func (db *DB) needLevelNCompactions() bool {
+	for idx := range db.levels {
+		if idx == 0 {
+			continue
+		}
+		if db.needLevelNCompaction(idx) {
 			db.compactionChan <- idx
-			readyToExit = false
+			return true
 		}
 	}
-	return readyToExit
+	return false
+}
+
+func (db *DB) needLevelNCompaction(level int) bool {
+	return db.levels[level].TotalSize() > calculateLevelSize(level)
```

Comment:
### Runtime Errors

- **Potential Index Out of Range**: In `needLevelNCompaction`, ensure that `level` is within the valid range of indices for `db.levels`. This can be done by adding a check:

```go
if level < 0 || level >= len(db.levels) {
    return false
}
```